sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py

```diff
@@ -0,0 +1,166 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.portability.api.beam_runner_api_pb2 import TestStreamPayload
+from apache_beam.utils import timestamp
+from apache_beam.utils.timestamp import Timestamp
+
+
+class StreamingCache(object):
+  """Abstraction that holds the logic for reading and writing to cache.
+  """
+  def __init__(self, readers):
+    self._readers = readers
+
+  class Reader(object):
+    """Abstraction that reads from PCollection readers.
+
+    This class is an Abstraction layer over multiple PCollection readers to be
+    used for supplying a TestStream service with events.
+
+    This class is also responsible for holding the state of the clock, injecting
+    clock advancement events, and watermark advancement events.
+    """
+    def __init__(self, readers):
+      # This timestamp is used as the monotonic clock to order events in the
+      # replay.
+      self._monotonic_clock = timestamp.Timestamp.of(0)
```

Comment:
MIN_TIMESTAMP?

Reply:
We do timestamp differences between timestamps in the _advance_processing_time method, which can overflow if this is MIN_TIMESTAMP. This isn't a Watermark, so it's okay to set this to 0 as we can't have an element with a wall time at 0.

Reply:
Cool, ty